### PR TITLE
Number format regardless of locale

### PR DESCRIPTION
**Pull request by TheBooper**
Stat values are converted to a string as part of building the statsd protocol string and creating the UDP payload. Coercion from float to string folows locale formats, which for some countries uses comma instead of period for floats.

E.x. 
```
echo 'US locale' . PHP_EOL;
$client->increment('test', 1, 'tag', 1000.001);
$client->increment('test', 1, 'tag', 1000.001);
$client->decrement('test', 1, 'tag', 1000.001);
// Set 
setlocale(LC_ALL, 'nl_NL');

echo 'Dutch locale' . PHP_EOL;
$client->increment('test', 1, 'tag', 1000.001);
$client->increment('test', 1, 'tag', 1000.001);
$client->decrement('test', 1, 'tag', 1000.001);
```
yields
```
US locale
Preformat:  1000.001
Postformat:  1000.00100
Preformat:  1000.001
Postformat:  1000.00100
Preformat:  -1000.001
Postformat:  -1000.00100
Dutch locale
Preformat:  1000,001
Postformat:  1000.00100
Preformat:  1000,001
Postformat:  1000.00100
Preformat:  -1000,001
Postformat:  -1000.00100
```
before and after using `number_format` when changing to a string.

### CODE REVIEW

File: src/DogStatsd.php [hunk 1]
```diff
@@ -555,4 +568,21 @@ private function eventUdp($vals)
...
+     *
+     * @return string Formatted value
+     */
+    protected function normalizeStat($value) {
+      // Controlls the way things are converted to a string.
+      // Otherwise localization settings impact float to string conversion (e.x 1.3 -> 1,3 and 10000 => 10,000)
+      // If one would rather control formatting themselves adn dont want to override the class, they can pass in a string
+      if (!is_float($value)) {
+        return $value;
+      }
```
Comment (hush-hush) [inline src/DogStatsd.php:585]: Could we always return a string (even for non float number) ?

Comment (TheBooper) [reply src/DogStatsd.php:585]: Sure, only floats have the formatting issue in php but for consistency we can handle all types like this. The only thing that really changes is that all types will be padded to whatever precision is configured. I.e. 2 => 2.00, 0.026 => 0.03

File: src/DogStatsd.php [hunk 2]
```diff
@@ -69,8 +73,9 @@ class DogStatsd
      * api_key and app_key
      *
      * @param array $config
+     * @param int   $decimal_precision
      */
-    public function __construct(array $config = array())
+    public function __construct(array $config = array(), $decimal_precision = 5)
```
[the remark this answers is not in the document]
Comment (TheBooper) [reply src/DogStatsd.php:78]: Yeah i'll change it to 2; if we dont support having this be configurable this takes away functionality from the client library since it seems DD support higher precision than 2 decimal places. I know internally, we would need to control precision.

[the remark this answers is not in the document]
Comment (hush-hush) [reply src/DogStatsd.php:78]: OK, then let's make that option part of the `$config` array to follow the same behavior that other option

File: src/DogStatsd.php [hunk 3]
```diff
@@ -555,4 +568,17 @@ private function eventUdp($vals)
...
+     * Normalize the value witout locale consideration before queuing the metric for sending
+     *
+     * @param string $value The value to normalize
+     *
+     * @return string Formatted value
+     */
+    protected function normalizeStat($value) {
+      // Controlls the way things are converted to a string.
+      // Otherwise localization settings impact float to string conversion (e.x 1.3 -> 1,3 and 10000 => 10,000)
+      return number_format($value, $this->decimalPrecision, '.', '');
```
Comment (hush-hush) [inline src/DogStatsd.php:582]: This force `.00` on all numbers and will increase the network usage for all payload for nothing. What do you think about trimming them ?

```php
return trim(number_format($value, $this->decimalPrecision, '.', ''), ".0");
```

Comment (TheBooper) [reply src/DogStatsd.php]: Needed a bit more logic than your snippet to trim, but should be all set now.

```
<?php

function normalizeStat($value) {
  // Controlls the way things are converted to a string.
  // Otherwise localization settings impact float to string conversion (e.x 1.3 -> 1,3 and 10000 => 10,000)
  $string_representation = rtrim(number_format($value, 5, '.', ''), "0");
  $last_char = substr($string_representation, -1, 1);
  if ($last_char == '.') {
    $string_representation = substr($string_representation, 0, strlen($string_representation) - 1);
  }
  return $string_representation;
}


echo normalizeStat(2.00) . PHP_EOL;
echo normalizeStat("96.000000") . PHP_EOL;
echo normalizeStat("2.006"). PHP_EOL;
echo normalizeStat(2.2222222222). PHP_EOL;
echo normalizeStat(2.01). PHP_EOL;
[rmadden@webphp-ftz2fns php-datadogstatsd]$ sudo php -f /tmp/normalizeTest.php
2
96
2.006
2.22222
2.01
```
Also tested for this in the tests i added, let me know if you need anything else.

Comment (TheBooper) [reply src/DogStatsd.php]: Yeah, trim's character class functionality has some edge case issues here since it chomps all consecutive characters of "0" or "." so 2000.00 becomes 2
```
TEST A
2
96
2.006
2.22222
2.01
2000
2000
TEST B
2
96
2.006
2.22222
2.01
2
2
[rmadden@webphp-ftz2fns php-datadogstatsd]$ cat /tmp/normalizeTest.php
<?php

function normalizeStatA($value) {
  // Controlls the way things are converted to a string.
  // Otherwise localization settings impact float to string conversion (e.x 1.3 -> 1,3 and 10000 => 10,000)
  $string_representation = rtrim(number_format($value, 5, '.', ''), "0");
  $last_char = substr($string_representation, -1, 1);
  if ($last_char == '.') {
    $string_representation = substr($string_representation, 0, strlen($string_representation) - 1);
  }
  return $string_representation;
}

function normalizeStatB($value) {
  // Controlls the way things are converted to a string.
  // Otherwise localization settings impact float to string conversion (e.x 1.3 -> 1,3 and 10000 => 10,000)
  return trim(number_format($value, 5, '.', ''), ".0");
}

echo "TEST A" . PHP_EOL;
echo normalizeStatA(2.00) . PHP_EOL;
echo normalizeStatA("96.000000") . PHP_EOL;
echo normalizeStatA("2.006"). PHP_EOL;
echo normalizeStatA(2.2222222222). PHP_EOL;
echo normalizeStatA(2.01). PHP_EOL;
echo normalizeStatA(2000). PHP_EOL;
echo normalizeStatA(2000.00). PHP_EOL;
echo "TEST B" . PHP_EOL;
echo normalizeStatB(2.00) . PHP_EOL;
echo normalizeStatB("96.000000") . PHP_EOL;
echo normalizeStatB("2.006"). PHP_EOL;
echo normalizeStatB(2.2222222222). PHP_EOL;
echo normalizeStatB(2.01). PHP_EOL;
echo normalizeStatB(2000). PHP_EOL;
echo normalizeStatB(2000.00). PHP_EOL;
```
Most concise way for me to show the difference between the functions

File: src/DogStatsd.php [hunk 4]
```diff
@@ -555,4 +568,23 @@ private function eventUdp($vals)
...
+     * Normalize the value witout locale consideration before queuing the metric for sending
+     *
+     * @param string $value The value to normalize
+     *
+     * @return string Formatted value
+     */
+    function normalizeStat($value) {
+      // Controlls the way things are converted to a string.
+      // Otherwise localization settings impact float to string conversion (e.x 1.3 -> 1,3 and 10000 => 10,000)
+      $string_representation = rtrim(number_format($value, 5, '.', ''), "0");
```
[the remark this answers is not in the document]
Comment (TheBooper) [reply src/DogStatsd.php:582]: Yep, good point, slipped my mind. Let me just run that through my own tests just to make certain.

**Comment by hush-hush** [comment]
Thanks a lot for this PR @TheBooper
